### PR TITLE
Remove globall calls of get_settings

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR is to remove all global calls of `get_settings()`

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

Current list of `get_settings` call
```
/home/czaki/.pyenv/versions/napari/bin/python -m napari
/home/czaki/Projekty/napari/napari/utils/_octree.py:77: RuntimeWarning: using settings in global context:
/home/czaki/Projekty/napari/napari/utils/config.py:55 # only async is enabled, not the octree.
octree_config = get_octree_config()

##########
  settings = get_settings()
/home/czaki/Projekty/napari/napari/utils/_octree.py:45: RuntimeWarning: using settings in global context:
/home/czaki/Projekty/napari/napari/utils/config.py:55 # only async is enabled, not the octree.
octree_config = get_octree_config()

##########
  async_var = get_settings().experimental.async_
/home/czaki/Projekty/napari/napari/plugins/_plugin_manager.py:104: RuntimeWarning: using settings in global context:
/home/czaki/Projekty/napari/napari/plugins/__init__.py:9 plugin_manager = NapariPluginManager()
plugin_manager._initialize()

##########
  plugin_settings = get_settings().plugins
/home/czaki/Projekty/napari/napari/plugins/__init__.py:11: RuntimeWarning: using settings in global context:
/home/czaki/Projekty/napari/napari/plugins/__init__.py:11 # Disable plugins listed as disabled in settings.
plugin_manager._blocked.update(get_settings().plugins.disabled_plugins)

##########
  plugin_manager._blocked.update(get_settings().plugins.disabled_plugins)
```

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by tlambert03** [comment]
can you leave a comment with a little more justification for calling `settings_module.get_settings()` instead of `from ... import ...; get_settings()`

note that a lot of this is slated to change when the settings context manager (work that didn't make it into #2932) is finished

**Comment by Czaki** [comment]
> can you leave a comment with a little more justification for calling `settings_module.get_settings()` instead of `from ... import ...; get_settings()`

because `__new__` method of `EventedMetaclass` calls field `default_factory` on class create and `_current_theme` call `get_settings`. So there is a need to check if `_SETTINGS` is None. So to avoid double module import also `get_settings` is called from module.

> note that a lot of this is slated to change when the settings context manager (work that didn't make it into #2932) is finished

I agree that this could help. And this change also need to solve this global calls. But I'm not sure how long we should wait for this.

**Comment by tlambert03** [comment]
> because __new__ method of EventedMetaclass calls field default_factory on class create and _current_theme call get_settings

ah right.  thanks.  that would be a very good thing to include in a comment in the code somewhere.  I suspect that would surprise most newcomers reading the code

File: napari/settings/__init__.py [hunk 1]
```diff
@@ -43,21 +44,35 @@ def get_settings(path=_NOT_SET) -> NapariSettings:
...
+    curframe = inspect.currentframe()
+    calframe = inspect.getouterframes(curframe, 2)
+    for frame in calframe[1:]:
+        if (
+            frame.function == "__new__"
+            and frame.filename
+            == '/home/czaki/Projekty/napari/napari/utils/events/evented_model.py'
+            and _SETTINGS is not None
+        ):
+            # contructor of EventedModel
```
Comment (Czaki) [inline napari/settings/__init__.py:61]: @tlambert03 this comment should be here or in another place?

**Comment by Czaki** [comment]
only octree and async left 
```
/home/czaki/Projekty/napari/napari/utils/_octree.py:77: RuntimeWarning: using settings in global context:
/home/czaki/Projekty/napari/napari/utils/config.py:55 # only async is enabled, not the octree.
octree_config = get_octree_config()

##########
  settings = get_settings()
/home/czaki/Projekty/napari/napari/utils/_octree.py:45: RuntimeWarning: using settings in global context:
/home/czaki/Projekty/napari/napari/utils/config.py:55 # only async is enabled, not the octree.
octree_config = get_octree_config()

##########
  async_var = get_settings().experimental.async_
/home/czaki/.pyenv/versions/napari/lib/python3.8/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: CUDA unknown error - this may be due to an incorrectly set up environment, e.g. changing env variable CUDA_VISIBLE_DEVICES after program start. Setting the available devices to be zero. (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:109.)
  return torch._C._cuda_getDeviceCount() > 0
```

**Comment by goanpeca** [comment]
Thanks @Czaki, trying now

**Comment by tlambert03** [comment]
kinda struggling with this PR.  This feels like a workaround solution to the problem.  If I understand correctly, the core problem is that, in partseg, you'd like to use your own `settings.yaml`, but currently there's a "race" to call `get_settings` first, is that corrrect?  I'd like to open an issue instead, and propose a refactor of `get_settings` that doesn't require us to race to call it first.  

see https://github.com/napari/napari/issues/3521

**Comment by Czaki** [comment]
It is not simply racing of `get_settings` call. It is problem, that `get_settings` is called on import time. Especially settings are used when importing the Image layer. Or a problem with pydantic which call `_current_theme` during build class (`__new__` call). 

The core of this PR was inspecting call in get_settings call which prevents of calling get settings on import call. 

So I think that This PR contains all required reactors about which you ask to be described. So I do not understand why not push this PR instead of writing a separate issue (I have a plan to close PR one by one)

**Comment by tlambert03** [comment]
> It is not simply racing of get_settings call. It is problem, that get_settings is called on import time.

This wouldn't _be_ a problem if you, and external program, could simply change the settings path by some later call.  The problem as I see it is that the path and contents of the settings can only be set in that very first call

**Comment by Czaki** [comment]
> This wouldn't _be_ a problem if you, and external program, could simply change the settings path by some later call. The problem as I see it is that the path and contents of the settings can only be set in that very first call

There is still a problem of testing code against older napari versions and instant dump of settings after the call of the constructor. With the current code, it could corrupt napari settings save (of default working one).

**Pull request closed without merging**